types/aws-lambda/trigger/cognito-user-pool-trigger/custom-email-sender.d.ts

```diff
@@ -0,0 +1,86 @@
+import { Handler } from '../../handler';
+import {
+    BaseTriggerEvent,
+    StringMap,
+} from './_common';
+
+export interface BaseCustomEmailSenderTriggerEvent<T extends string> extends BaseTriggerEvent<T> {
+    request: {
```

Comment:
Source of the `request` type?

Reply:
I didn't find anything about these types in official docs, because this feature was added recently.
So I decided to debug all these events by myself and this is what  got:

P.S.: object consists of `[key]: value`, where `key` is like `${trigger type}_${trigger event type}_${trigger event  sub type if exists}`, for example:  `CustomEmailSender_AccountTakeOverNotification_BLOCK` or `CustomEmailSender_SignUp`, and value is event object that I recieved

```
{
  "CustomEmailSender_AccountTakeOverNotification_BLOCK": {
    "version": "1",
    "triggerSource": "CustomEmailSender_AccountTakeOverNotification",
    "region": "${region}",
    "userPoolId": "${userPoolId}",
    "userName": "${userName}",
    "callerContext": {
      "awsSdkVersion": null | string,
      "clientId": null | string
    },
    "request": {
      "type": "customEmailSenderRequestV1",
      "code": null,
      "clientMetadata": null,
      "userAttributes": {
        "LOGIN_TIME": "Friday, March 19, 2021 02:25 PM UTC",
        "CITY": "${CITY} as string" | null,
        "COUNTRY": "${COUNTRY} as string" | null,
        "ACCOUNT_TAKE_OVER_ACTION": "BLOCK",
        "FEEDBACK_TOKEN": "${FEEDBACK_TOKEN}",
        "ONE_CLICK_LINK_VALID": "${ONE_CLICK_LINK_VALID}",
        "DEVICE_NAME": "(Opera 74, Linux) as string" | null,
        "EVENT_ID": "${EVENT_ID}",
        "IP_ADDRESS": "${IP_ADDRESS}",
        "USER_NAME": "${USER_NAME}",
        "ONE_CLICK_LINK_INVALID": "${ONE_CLICK_LINK_INVALID}"
      }
    }
  },
  "CustomEmailSender_AccountTakeOverNotification_NO_ACTION": {
    "version": "1",
    "triggerSource": "CustomEmailSender_AccountTakeOverNotification",
    "region": "${region}",
    "userPoolId": "${userPoolId}",
    "userName": "${userName}",
    "callerContext": {
      "awsSdkVersion": null | string,
      "clientId": null | string
    },
    "request": {
      "type": "customEmailSenderRequestV1",
      "code": null,
      "clientMetadata": null,
      "userAttributes": {
        "LOGIN_TIME": "Friday, March 19, 2021 02:44 PM UTC",
        "CITY":  "${CITY} as string" | null,
        "COUNTRY": "${COUNTRY} as string" | null,
        "ACCOUNT_TAKE_OVER_ACTION": "NO_ACTION",
        "ONE_CLICK_LINK_VALID": "${ONE_CLICK_LINK_VALID}",
        "FEEDBACK_TOKEN": "${FEEDBACK_TOKEN}",
        "DEVICE_NAME": "${DEVICE_NAME} as string" | null,
        "EVENT_ID": "${EVENT_ID}",
        "IP_ADDRESS": "${IP_ADDRESS}",
        "USER_NAME": "${USER_NAME}",
        "ONE_CLICK_LINK_INVALID": "${ONE_CLICK_LINK_INVALID}"
      }
    }
  },
  "CustomEmailSender_AccountTakeOverNotification_MFA": {
    "version": "1",
    "triggerSource": "CustomEmailSender_AccountTakeOverNotification",
    "region": "${region}",
    "userPoolId": "${userPoolId}",
    "userName": "${userName}",
    "callerContext": {
      "awsSdkVersion": null | string,
      "clientId": null | string
    },
    "request": {
      "type": "customEmailSenderRequestV1",
      "code": null,
      "clientMetadata": null,
      "userAttributes": {
        "LOGIN_TIME": "Friday, March 19, 2021 09:54 AM UTC",
        "CITY":  "${CITY} as string" | null,
        "COUNTRY": "${COUNTRY} as string" | null,
        "ACCOUNT_TAKE_OVER_ACTION": "MFA",
        "ONE_CLICK_LINK_VALID": "${ONE_CLICK_LINK_VALID}",
        "FEEDBACK_TOKEN": "${FEEDBACK_TOKEN}",
        "DEVICE_NAME": "${DEVICE_NAME} as string" | null,
        "EVENT_ID": "${EVENT_ID}",
        "IP_ADDRESS": "${IP_ADDRESS}",
        "USER_NAME": "${USER_NAME}",
        "ONE_CLICK_LINK_INVALID": "${ONE_CLICK_LINK_INVALID}"
      }
    }
  },
  "CustomEmailSender_AccountTakeOverNotification_MFA_IF_CONFIGURED": {
    "version": "1",
    "triggerSource": "CustomEmailSender_AccountTakeOverNotification",
    "region": "${region}",
    "userPoolId": "${userPoolId}",
    "userName": "${userName}",
    "callerContext": {
      "awsSdkVersion": null | string,
      "clientId": null | string
    },
    "request": {
      "type": "customEmailSenderRequestV1",
      "code": null,
      "clientMetadata": null,
      "userAttributes": {
        "LOGIN_TIME": "Friday, March 19, 2021 09:49 AM UTC",
        "CITY":  "${CITY} as string" | null,
        "COUNTRY": "${COUNTRY} as string" | null,
        "ACCOUNT_TAKE_OVER_ACTION": "MFA_IF_CONFIGURED",
        "ONE_CLICK_LINK_VALID": "${ONE_CLICK_LINK_VALID}",
        "FEEDBACK_TOKEN": "${FEEDBACK_TOKEN}",
        "DEVICE_NAME": "${DEVICE_NAME} as string" | null,
        "EVENT_ID": "${EVENT_ID}",
        "IP_ADDRESS": "${IP_ADDRESS}",
        "USER_NAME": "${USER_NAME}",
        "ONE_CLICK_LINK_INVALID": "${ONE_CLICK_LINK_INVALID}"
      }
    }
  },
  "CustomEmailSender_AccountTakeOverNotification_MFA_REQUIRED": {
    "version": "1",
    "triggerSource": "CustomEmailSender_AccountTakeOverNotification",
    "region": "${region}",
    "userPoolId": "${userPoolId}",
    "userName": "${userName}",
    "callerContext": {
      "awsSdkVersion": null | string,
      "clientId": null | string
    },
    "request": {
      "type": "customEmailSenderRequestV1",
      "code": null,
      "clientMetadata": null,
      "userAttributes": {
        "LOGIN_TIME": "Friday, March 19, 2021 09:54 AM UTC",
        "CITY":  "${CITY} as string" | null,
        "COUNTRY": "${COUNTRY} as string" | null,
        "ACCOUNT_TAKE_OVER_ACTION": "MFA_REQUIRED",
        "ONE_CLICK_LINK_VALID": "${ONE_CLICK_LINK_VALID}",
        "FEEDBACK_TOKEN": "${FEEDBACK_TOKEN}",
        "DEVICE_NAME": "${DEVICE_NAME} as string" | null,
        "EVENT_ID": "${EVENT_ID}",
        "IP_ADDRESS": "${IP_ADDRESS}",
        "USER_NAME": "${USER_NAME}",
        "ONE_CLICK_LINK_INVALID": "${ONE_CLICK_LINK_INVALID}"
      }
    }
  },
  "CustomEmailSender_SignUp": {
    "version": "1",
    "triggerSource": "CustomEmailSender_SignUp",
    "region": "${region}",
    "userPoolId": "${userPoolId}",
    "userName": "${userName}",
    "callerContext": {
      "awsSdkVersion": null | string,
      "clientId": null | string
    },
    "request": {
      "type": "customEmailSenderRequestV1",
      "code": "${encrypted_code}",
      "clientMetadata": null,
      "userAttributes": {
        "sub": "${sub}",
        "cognito:user_status": "UNCONFIRMED",
        "email_verified": "false",
        "cognito:email_alias": "${cognito:email_alias}",
        "custom:email_consent": "1",
        "email": "${email}"
      }
    }
  },
  "CustomEmailSender_UpdateUserAttribute": {
    "version": "1",
    "triggerSource": "CustomEmailSender_UpdateUserAttribute",
    "region": "${region}",
    "userPoolId": "${userPoolId}",
    "userName": "${userName}",
    "callerContext": {
      "awsSdkVersion": null | string,
      "clientId": null | string
    },
    "request": {
      "type": "customEmailSenderRequestV1",
      "code": "${encrypted_code}",
      "clientMetadata": null,
      "userAttributes": {
        "sub": "${sub}",
        "cognito:user_status": "CONFIRMED",
        "email_verified": "false",
        "email": "${email}"
      }
    }
  },
  "CustomEmailSender_ResendCode": {
    "version": "1",
    "triggerSource": "CustomEmailSender_ResendCode",
    "region": "${region}",
    "userPoolId": "${userPoolId}",
    "userName": "${userName}",
    "callerContext": {
      "awsSdkVersion": null | string,
      "clientId": null | string
    },
    "request": {
      "type": "customEmailSenderRequestV1",
      "code": "${encrypted_code}",
      "clientMetadata": null,
      "userAttributes": {
        "sub": "${sub}",
        "email_verified": "false",
        "cognito:user_status": "UNCONFIRMED",
        "cognito:email_alias": "${cognito:email_alias}",
        "custom:email_consent": "1",
        "email": "${email}"
      }
    }
  },
  "CustomEmailSender_ForgotPassword": {
    "version": "1",
    "triggerSource": "CustomEmailSender_ForgotPassword",
    "region": "${region}",
    "userPoolId": "${userPoolId}",
    "userName": "${userName}",
    "callerContext": {
      "awsSdkVersion": null | string,
      "clientId": null | string
    },
    "request": {
      "type": "customEmailSenderRequestV1",
      "code": "${encrypted_code}",
      "clientMetadata": {
        "custom_key": "custom_value"
      },
      "userAttributes": {
        "sub": "${sub}",
        "email_verified": "true",
        "cognito:user_status": "CONFIRMED",
        "cognito:email_alias": "${cognito:email_alias}",
        "email": "${email}"
      }
    }
  }
}

```